### PR TITLE
added Mapper methods and granularity for marshalling

### PR DESCRIPTION
**Pull request by Groostav**
Simplified my previous implementation such that it doesn't use enums and doesn't modify any public facing API's.

To properly preserve backwards compatibility, my new method must delegate to the extra-arg free one, such that the chain of wrappers, on a call to the new method, also includes calling the old method. Since the method is looking for any positive result (ie, once anybody returns true in any context we're done), the solution to this was simply to have

``` Java
public boolean isImmutableType(Class typeName, boolean includeBackwardsCompatible){
    return isImmutableType(typeName) || wrapped.isImmutableType(typeName, includeBackwardsCompatible)
}
```

This has the effect of invoking any user-written backwards-compatible-agnostic code to look for a result, and then look through the defaults in the proper scope.

The one issue I have with this implementation is that the performance of the `Mapper.isImmutableType(Class, boolean)` is porportional to the _square_ of the number of wrappers.

For example, if we have:
- StringeyMapper extends MapperWrapper 
- DomainishMapper extends MapperWrapper
- ImmutableMapper extends MapperWrapper

and we get them wrapped in the sequence

`StringeyMapper` delegates to `DomainishMapper` delegates to `ImmutableMapper` delegates to `defaultMapper`

then the call stack would be

| caller | call-site |
| --- | --- |
| _consumer_ | `stringyMapper.isImmutableType(t, true)` |
| stringyMapper | `stringyMapper.isImmutableType(t)` |
| stringyMapper | `domainishMapper.isImmutableType(t)` |
| domainishMapper | `immutableMapper.isImmutableType(t)` |
| immutableMapper | `defaultMapper.isImmutableType(t)` |
| immutableMapper | 1-arg method on defaultMapper returns |
| domainishMapper | 1-arg method on immutableMapper returns |
| stringyMapper | 1-arg method on domainishMapper returns |
| stringyMapper | `domainishMapper.isImmutableType(t, true)` |
| domainishMapper | **redundantly** `domainishMapper.isImmutableType(t)` |
| domainishMapper | **redundantly** `immutableMapper.isImmutableType(t)` |
| immutableMapper | **redundantly** `defaultMapper.isImmutableType(t)` |
| immutableMapper | 1-arg method on defaultMapper returns |
| domainishMapper | 1-arg method on immutableMapper returns |
| domainishMapper | `immutableMapper.isImmutableType(t, true)` |
| immutableMapper | **redundantly** `immutableMapper.isImmutableType(t)` |
| immutableMapper | **redundantly** `defaultMapper.isImmutableType(t)` |
| immutableMapper | 1-arg method on defaultMapper returns |
| immutableMapper | defaultMapper.isImmutableType(t, true); |
| immutableMapper | 2-arg method on defaultMapper returns |
| domainishMapper | 2-arg method on immutableMapper returns |
| stringeyMapper | 2-arg method on domainishMapper returns |
| _consumer_ | 2-arg method on stringeyMapper returns |

If I'm not mistaken this should be equal to numWrappers ^2  /2, but 13 != 8 so I might be mistaken.

This means that any user-written code for Types must be re-entrant, which was not a requirement before (though no sane implementation wouldn't be). Further it means a user that puts a break-point inside his `isIMmutableType` overload is going to see it hit _alot_ more frequently, though that doesn't seem like too much of a concern to me.

A bigger concern to me was performance: all those extra calls might make a negative impact on performance. To test this I created an array list of 1,000,000 ints, and implemented a solution that kept a map of currently-inspected types which it referenced, so as to remove the redundant calls. This implementation was slightly slower than the implementation that had the redundant calls, so performance appears not to be an issue, unless I made an error in writing the performance test.

using the test

``` Java
        ArrayList<Integer> ints = new ArrayList<>();
        for(int i = 0; i < 1_000_000; i++){
            ints.add(i);
        }

        StopWatch watch = new StopWatch();
        watch.start();
        String document = xstream.toXML(ints);
        System.out.println("took " + watch.getTime() + "ms to serialize");
        watch.stop();
        watch.reset();
        watch.start();

        Object result = xstream.fromXML(document);
        System.out.print("took " + watch.getTime() + "ms to deserialize");
```

and the following to address the problem

``` Java
//as a static field --ThreadLocal's have been around since java 1.2, and anonymous classes since 1.1
private static final ThreadLocal<Set<Class>> inspectingTypes = new ThreadLocal<Set<Class>>(){
    protected Set<Class> initialValue() { return new HashSet<>(); }
};

//in the isImmutableType(Class, boolean) method
Set<Class> classCallStack = inspectingTypes.get();

if( ! classCallStack.contains(type)) try {
    classCallStack.add(type);

    return isImmutableValueType(type)
            || wrapped.isImmutableValueType(type, includeBackwardsCompatibleTypes);
}
finally{
    classCallStack.remove(type);
}
else{
    return wrapped.isImmutableValueType(type, includeBackwardsCompatibleTypes);
}
```

I got

| scenario | serialization time | deserialization time |
| --- | --- | --- |
| with the problem | 931ms | 1720ms |
| without the problem | 928ms | 1672ms |
| with the wierd stack-set fix | 948ms | 1722ms |

which to me means its all a wash, thus probably just accepting these extra calls is OK.

### CODE REVIEW

**Comment by Groostav** [comment]
tagging @joehni for review!

File: xstream/src/java/com/thoughtworks/xstream/XStream.java [hunk 1]
```diff
@@ -11,106 +11,15 @@
...
 import com.thoughtworks.xstream.converters.reflection.ReflectionProvider;
 import com.thoughtworks.xstream.converters.reflection.SerializableConverter;
-import com.thoughtworks.xstream.core.ClassLoaderReference;
-import com.thoughtworks.xstream.core.DefaultConverterLookup;
-import com.thoughtworks.xstream.core.JVM;
-import com.thoughtworks.xstream.core.MapBackedDataHolder;
-import com.thoughtworks.xstream.core.ReferenceByIdMarshallingStrategy;
-import com.thoughtworks.xstream.core.ReferenceByXPathMarshallingStrategy;
-import com.thoughtworks.xstream.core.TreeMarshallingStrategy;
+import com.thoughtworks.xstream.core.*;
```
Comment (Groostav) [inline xstream/src/java/com/thoughtworks/xstream/XStream.java]: sorry about this, my IDE did this automatically, I'll revert it when I get a chance

File: xstream/src/java/com/thoughtworks/xstream/mapper/ImmutableTypesMapper.java [hunk 2]
```diff
@@ -22,22 +22,41 @@
...
+    @Override
+    public boolean isImmutableValueType(Class<?> type, boolean includeBackwardsCompatibleTypes) {
+        return pathRetentionByType.containsKey(type)
+                ? isImmutableType(type, includeBackwardsCompatibleTypes ? RETAIN_FOR_COMPATIBILITY : RETAIN_NEVER)
+                : super.isImmutableValueType(type, includeBackwardsCompatibleTypes);
+    }
+
+    private boolean isImmutableType(Class<?> type, int targetLevel) {
+        int retentionPolicy = pathRetentionByType.get(type);
+        return retentionPolicy == RETAIN_NEVER || retentionPolicy == RETAIN_ALWAYS;
```
Comment (Groostav) [inline xstream/src/java/com/thoughtworks/xstream/mapper/ImmutableTypesMapper.java]: now _thats_ both odd and scary... how does this not fail a test?

**Comment by Groostav** [comment]
I've completely screwed this up. Need to work on my boolean algebra.

**Comment by Groostav** [comment]
OK, so I added a test to proove that the unmarshalling class does keep references when in backwards compatible mode, which was the problem I believe I was having.

**Comment by joehni** [comment]
Hi Geoff, IMHO it is still too complicated. Let's reiterate for a moment:

Basically we want to avoid the memory consuming references to the unmarshalled objects. Nothing changes while marshalling. The only problem is that XML written with older versions of XStream may contain such a reference and cannot be unmarshalled anymore. Therefore we need a configuration, which types have to be referenced at unmarshalling time.

The only question of the unmarshaller is now, do I have to keep a reference? The only way to access the configuration is the Mapper, so it requires a (new) method `isReferenceable(Class type)`.
With the `ImmutableMapper` you've found the best place to keep this information. And we need a way to set the flag, e.g. by adding an overloaded `addImmutableType(Class type, boolean referenceable)` to the XStream facade.

I do not see any necessity that the new method of the mapper is connected with the method isImmutableType. And the only one requesting the information is currently the `AbstractReferenceUnmarshaller`. All we have to do is a default for `XStream.addImmutableType` with the flag set to true (i.e. type is referenceable) and add any type in setupImmutableTypes that was already declared immutable in XStream version < 1.4.0 with the flag set to false.

BTW: XStream master requires Java 1.6. So you have to keep all the generics and `@Override` annotations. They can be backported in a binary compatible way for the 1.4.x branch. That is just not possible for an enum.

Regarding the benchmark: Something like `StopWatch` is useless for code that is executed within fractions of milliseconds. Look at [JMH](http://openjdk.java.net/projects/code-tools/jmh/) if you want to have useful results.

**Comment by Groostav** [comment]
The issue with adding an `isReferentiable` method is that if somebody has a complex behavior for `isImmutable` that they expect to have executed on deserialization, it wont be.

In our case, even a behavior as simple as 

```
@Override
public boolean isImmutableType(Class type){
    return type.isAnnotationPresent(com.myCompany.ImmutableAnnotation.class)
        || super.isImmutable(type);
}
```

With this strategy users would then have to add

```
@Override
public boolean isReferentiable(Class type){
    return isImmutableType(type);
}
```

which also doesn't _quite_ get the right behaviour since this will hit some mid-point in the wrapper chain.

And that's a surprise I wouldn't want to expose anyone too, and is effectively what I'm asking the library to do by default.

That said, whats more unexpected is that forking behaviour I mentioned above.

**Comment by Groostav** [comment]
actually I'm being really silly, from a boolean algebra view this is quite simple.

Will do.

**Comment by Groostav** [comment]
Alright, so I think I'm done tell me what you think @joehni.

I dont know if github has a way to generate diffs, but I can make them easily enough and send them to you on the mailing list easily enough.

I'm not sure whats involved in your back porting project, but I figured if you could tolerate some `@Override` annotations then you would be able to tolerate all the methods on the `MapperWrapper` being tagged with them, since I hate the idea of 2 of the overriding methods being tagged with it and the rest not being tagged with it.

**Comment by Groostav** [comment]
ahah:

regarding generating diffs and patches, checkout [thisPR.patch](./2.patch) and [thisPR.diff](./2.diff)

nifty stuff github...

**Comment by Groostav** [comment]
@joehni thanks for your patience in dealing with my bumbling around an obvious solution. Are you going to apply this to the latest build (and can I close this issue?)

**Comment by joehni** [comment]
Hi Geoff,

when I look at your diff, then I wonder, what did you use as parent for your XStream fork? It does not seem to be the current code base... ?!?

**Comment by Groostav** [comment]
I  downloaded source + binaries from [XStream/downloads "source distribution"](http://xstream.codehaus.org/download.html) of the stable version

**Comment by Groostav** [comment]
I'll rebase that stuff into [a proper fork](../../xstream)

**Pull request closed without merging**